Add support for mercurial DVCS

mbezjak opened the pull request:
Mercurial is very similar to git, so I based `HgVcsSupport.java` on existing `GitVcsSupport.java`.

I tried all relevant pride commands (update, reinit, remove, list, do, export). Everything works as expected (rebasing, pull + update, activating, etc.).

However, this is still a work in progress. I'd still like to:
- [x] create integration tests for hg that use pride from cmdline (like user would); this should test:
  - [x] pride init
  - [x] clone without mirror
  - [x] clone with mirror
  - [x] update module when it's clean (--no-repo-cache)
  - [x] update module when it's clean (--use-repo-cache)
  - [x] update module when it's dirty
  - [x] update module when it has unpublished commits
  - [x] update module when it has merge conflicts
  - [x] `pride do -- hg status` should work as expected
  - [x] `pride list` should work as expected
  - [x] remove code duplications (e.g. writing to .hg/hgrc)
- [x] squash all those commits into one (to remove useless commits)
- [x] mention mercurial in README.md

Commit:
Support mercurial DVCS.

Mercurial is very similar to git, `HgVcsSupport.java` is based on
`GitVcsSupport.java`. All relevant pride commands (update, reinit,
remove, list, do, export) work as expected. Tests cover most of
`HgVcsSupport` functionality:

- with and without repo cache
- pristine repo, with uncommitted changes, with unpublished changes
- running pride: init, update, list, do -- hg status

In case of conflicting changes: fail the merge and let the user
handle it on it's own.

refs:
- https://www.selenic.com/hg/help/merge-tools

lptr commented:
This is great. For the integration tests you can use `AbstractIntegrationSpec`. You might want to add some relevant cases to the existing integration tests. For example, `PrideInitIntegrationTest` could add a Mercurial module as well. `hg` should be installed on Travis CI, so you don't need anything extra there.

mbezjak commented:
Done!

Mercurial support works as expected. The only thing I had to change is [RepoCache](https://github.com/prezi/pride/pull/158/files#diff-040f2620242d491757f6c00a046d9cf5R48) to handle unrelated repos. Everything else is based on GitVcsSupport.

Tests cover most of HgVcsSupport functionality:
- with and without repo cache
- pristine repo, with uncommited changes, with unpublished changes
- running pride: init, update, list, do -- hg status

In addition, those tests should also be fairly readable.

Tomorrow, I'll rebase and squash all those commits to just one. Afterwards, if you have no objections, I'd consider that to be merge-worthy.

Comments welcome.

mbezjak commented:
Rebased and squashed, but I just remembered that I should mention mercurial in README.md. I'll do that tomorrow.

mbezjak commented:
Done. Good to merge, I think.

lptr commented:
Hey. Thanks. I'm currently not able to check this out more thoroughly, but
next week I will.
On Wed, 21 Oct 2015 at 11:12, Miro Bezjak <notifications@github.com> wrote:

> Done. Good to merge, I think.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/prezi/pride/pull/158#issuecomment-149829457>.
>
-- 
-- 
Lóránt

mbezjak commented:
Ok. No problem.

lptr commented on README.md:
```diff
@@ -47,6 +47,9 @@ pride add network-component backend-api
 
+# To add a mercurial module, use --repo-type
+pride add --repo-type hg mercurial-project
+
```
`HgVcsSupportFactory.canSupport()` should detect this automatically.

mbezjak replied:
As far as I can see, `HgVcsSupportFactory.canSupport()` is only used in [InitAction](https://github.com/prezi/pride/blob/master/pride/src/main/java/com/prezi/pride/cli/commands/actions/InitAction.java#L83). But this relates to `AddAction`. I think the problem is in
[AddCommand.getModulesToAdd](https://github.com/prezi/pride/blob/master/pride/src/main/java/com/prezi/pride/cli/commands/AddCommand.java#L71). It picks up `VcsSupport` based on either an explicit repo type or a default one (git).

I was going to solve that one in the next PR. But I can try to solve it here if you want. It would probably mean changing `AddCommand`.

lptr commented on README.md:
```diff
@@ -118,6 +121,10 @@ To add modules by cloning them:
+To clone mercurial module:
+
+    $ pride add --repo-type hg <repo>
+
```
Same here: `HgVcsSupportFactory.canSupport()` should detect this automatically.

lptr commented on pride-core/src/main/java/com/prezi/pride/vcs/RepoCache.java:
```diff
@@ -42,7 +45,17 @@ public void checkoutThroughCache(VcsSupport vcsSupport, final String repositoryU
+				log.info("Cleaning up stale repo ({}) from the cache: {}", repositoryUrl, moduleInCacheName);
+				FileUtils.deleteDirectory(moduleInCache);
+				vcsSupport.checkout(repositoryUrl, moduleInCache, null, false, true);
+			}
```
I'd like to avoid relying on exceptions here. What is the exact condition you are handling here? How can we work around it, or handle it more gracefully than indiscriminately catching all `PrideException`s?

mbezjak replied:
The exact condition is tested in `HgCacheRepoTest.groovy`. Basically what if
repo in cache and the one that you are adding become unrelated? Then the cache
should be removed first.

I tried googling if there is way to detect upfront if repos are unrelated. No
luck. As far as I can tell, you just need try and see if it'll fail or not.

I'll think about this one a little bit. Maybe there is a better way.

lptr commented on pride-hg-support/src/main/java/com/prezi/pride/vcs/hg/HgVcsSupport.java:
```diff
@@ -0,0 +1,210 @@
+		} catch (PrideException e) {
+			// `hg outgoing` exits (stupidly) with status code 1 when there are no outgoing changes
+			return false;
+		}
```
We should handle this better. Perhaps refactor `ProcessUtils.executeIn()` in a way that it's able to return the exit value when needed (and handle the error as it does now in other cases). Maybe do that in a separate pull-request first.

lptr commented on pride-hg-support/src/main/java/com/prezi/pride/vcs/hg/HgVcsSupport.java:
```diff
@@ -0,0 +1,210 @@
+		File hgrcFile = new File(targetDirectory, ".hg/hgrc");
+		String hgrcText = FileUtils.readFileToString(hgrcFile, Charsets.UTF_8);
+		String activated = hgrcText.replaceAll("default = .*", "default = " + repositoryUrl);
+		FileUtils.write(hgrcFile, activated, Charsets.UTF_8);
```
Can you explain what's happening here? Why do we need to manually edit a config file? Can we do this via some Mercurial command instead? Say, `hg config`?

mbezjak replied:
Nope. I googled.

refs:
- http://stackoverflow.com/questions/4659135/can-i-set-mercurial-config-options-programatically

lptr commented on pride/src/integTest/data/PrideInitIntegrationTest/git-module/build.gradle:
```diff
@@ -17,7 +17,7 @@ configurations {
 }
 
-dynamicDependencies {
+dependencies {
```
This was intentionally left as `dynamicDependencies` so that there's at least one test that exercises the backwards compatibility support. That said, it should probably be a separate integration test. In any case, please revert it for this pull-request.

lptr commented on pride/src/integTest/groovy/com/prezi/pride/test/HgCacheRepoTest.groovy:
```diff
@@ -0,0 +1,34 @@
+	private void ensurePrideModuleIsCached(String moduleName) {
+		def tempWorkingDir = file('temporary-pride-working-directory-used-to-init-cache')
+		initializePrideWithModule(cache: true, tempWorkingDir, moduleName)
+		assert tempWorkingDir.deleteDir()
```
This should assert that the directory exists, and delete it in a `finally { }` block.

lptr commented on pride/src/integTest/groovy/com/prezi/pride/test/HgSupport.groovy:
```diff
@@ -0,0 +1,56 @@
+package com.prezi.pride.test
+
+trait HgSupport {
```
Nice. Let's call this `HgTestSupport` to signify that it's related to tests.

lptr commented on pride/src/integTest/groovy/com/prezi/pride/test/HgSupport.groovy:
```diff
@@ -0,0 +1,56 @@
+    assert file(workingDir).mkdir()
+    pride workingDir: workingDir, 'init'
+    prideAdd opts, workingDir, moduleName
+  }
```
Can this be part of `AbstractIntegrationSpec`?

mbezjak replied:
It can with a `repoType` option. Or better yet, no option at all if we change
`AddCommand` (see comment above for `README.md`).

mbezjak commented:
Thanks for the comments!

I currently have an approaching deadline for a project so I'm not sure when I'll have the time to work on this. I'll try to work 20 minutes each day. We'll see how long it'll take me.

Here is the current checklist:
- [ ] revert change to `dynamicDependencies`
- [ ] `HgSupport` -> `HgTestSupport`
- [ ] `HgCacheRepoTest`: assert dir exists + delete in finally
- [ ] ProcessUtils.executeIn should be able to return exit value when needed (`hasUnpublishedChanges`) (new PR)
- [ ] eliminate the need for `--repo-type hg`
- [ ] `prideAdd` and `initializePrideWithModule` to `AbstractIntegrationSpec`
- [ ] is there a better way to handle unrelated repos?

mbezjak commented:
Just to be clear. I'm waiting for #161 to be merged so I can commit what I have currently before I tackle `--repo-type hg`. No rush; just wanted to let you know.

I've been thinking about how to eliminate the need for `--repo-type hg` when doing `pride add`. The only thing that seems valid to me is `for` loop over registered `*VcsSupport` + clone in `try/catch`. The main reason for this is that just from the URL (e.g. `http://repo-server/project`), one cannot know if this is `git` or `hg` project. One just needs to try to clone with git. If that fails then it could be a hg project and try that. Can you think of a better solution?

Crazyjavahacking commented:
Can we merge this please?

mbezjak commented:
Hi all.

Since gradle 3.1 introduced [composite builds], I'm no longer relying on pride. So I just wanted to let you know that I won't complete the above todos (most significant is to eliminate the need for `--repo-type hg`). Nevertheless, I've squashed and rebased the branch so that merge could be painless if you think that the current implementation is good enough (it has a bunch of passing tests but has a `--repo-type hg` quirk that does not affect git usage).

Kind regards

[composite builds]: https://docs.gradle.org/3.1/userguide/composite_builds.html

rosadam commented:
@mbezjak thanks for the update, I will review it.
